presto-main/src/main/java/com/facebook/presto/sql/analyzer/MaterializedViewInformationExtractor.java

```diff
@@ -0,0 +1,149 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.sql.analyzer;
+
+import com.facebook.presto.sql.tree.ComparisonExpression;
+import com.facebook.presto.sql.tree.DefaultTraversalVisitor;
+import com.facebook.presto.sql.tree.Expression;
+import com.facebook.presto.sql.tree.GroupBy;
+import com.facebook.presto.sql.tree.GroupingElement;
+import com.facebook.presto.sql.tree.Identifier;
+import com.facebook.presto.sql.tree.LogicalBinaryExpression;
+import com.facebook.presto.sql.tree.QuerySpecification;
+import com.facebook.presto.sql.tree.Relation;
+import com.facebook.presto.sql.tree.SingleColumn;
+import com.facebook.presto.sql.tree.Table;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+
+public class MaterializedViewInformationExtractor
+        extends DefaultTraversalVisitor<Void, MaterializedViewInformationExtractor.MaterializedViewInfo>
+{
+    @Override
+    protected Void visitQuerySpecification(QuerySpecification node, MaterializedViewInfo materializedViewInfo)
+    {
+        super.visitQuerySpecification(node, materializedViewInfo);
+        if (node.getLimit().isPresent()) {
+            materializedViewInfo.setLimitClausePresented(true);
+        }
+        return null;
+    }
+
+    @Override
+    protected Void visitRelation(Relation node, MaterializedViewInfo materializedViewInfo)
+    {
+        // Currently it only handles Table instance cases (Do not handle Join, AliasedRelation, etc)
+        if (node instanceof Table) {
```

Comment:
Throw exception if the node is not Table type.

Reply:
Could I throw it in the query optimizer (at the start of visitQuery I check if the base table is presented). Since the information extractor is called in the constructor of queryOptimizer, it will be better if we throw and catch the exception during the rewrite process.

Reply:
It's not same isn't it? If the MV and baseTables combination is not good, why wait till the query needs to be processed?